Implements user impersonation and the ability to set a custom date.

hcraT opened the pull request:
This patch implements user impersonation and dates preservation when importing the messages to Zulip.
Options are implemented to:
- enable/disable user impersonation
- allow zulip to create new users for missing ones
- define a map from elog users to zulip users
- enable/disable date preservation

hcraT committed:
Implements user impersonation and the ability to set a custom date.

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -4,7 +4,10 @@
+import sys
 import warnings
+import datetime
+import csv
```
nitpick: import order

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -57,19 +60,63 @@ def __init__(self, config, dry_run=False):
-            self.zulip = zulip.Client(config_file=config['zulip-rc'])
+            self.zulip = zulip.Client(
+                    config_file=config['zulip-rc'],
+                    insecure=True,
```
what does this do?
I suspect we should set this only under certain conditions?

hcraT replied:
Good find!. It is indeed a remain from local tests (the local server don't have a valid certificate).

hcraT replied:
I've also added an option that by default is false so that, if necessary, one can use this on insecure servers.

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -57,19 +60,63 @@ def __init__(self, config, dry_run=False):
+            self.zulip = zulip.Client(
+                    config_file=config['zulip-rc'],
+                    insecure=True,
+                    client='jabber_mirror' if self.impersonate else None
```
maybe a small comment here would also help understand what this means.

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -57,19 +60,63 @@ def __init__(self, config, dry_run=False):
             self._db = dataset.connect(config['database'])
             self.entry = self._db[self.table]
 
+            if not can_create_users:
+                response = self.zulip.get_members()
+                if response['result'] == 'success':
+                    self._existing_users = { m['email'] : m for m in response['members'] }
+                else:
+                    log.error(f'Error while querying the users: {response["msg"]}')
+                    sys.exit(1)
```
I'd move this block below in the `if self.impersonate` branch as this seems to be only used here.
`self._existing_users` doesn't have to be a instance attribute as it's used only in this method.

hcraT replied:
You are right about the `self._existing_users` and I will move it to the local scope. This block is here so that I don't need to add the `get_members()` method to the mock class. But if you prefer I can create a mock method that returns an empty response.

hcraT replied:
In the end I went for a different route. I moved all this code in the non-dry-run branch. In the dry-run I don't have access to the members of the Zulip server anyway so it doesn't make sense to validate the users.

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -57,19 +60,63 @@ def __init__(self, config, dry_run=False):
+        if self.impersonate:
+            if users_map_path:
+                with open(users_map_path, newline='') as f:
+                    data = list(csv.reader(f))
```
maybe document somewhere the expected format of that csv file.

hcraT replied:
You are right! I'll do it

hcraT replied:
I've added a check that the csv has the right fields with an error message if the fields are not present. Do you think that this is sufficient?

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -137,12 +184,32 @@ def _send_message(self, message, topic):
+        sender = None
+        if self.impersonate and self._users_map:
+            a = attributes['Author']
+            sender = self.zulip.get_profile()['email']
+            if a in self._users_map:
+                sender = self._users_map[a]['email']
```
```suggestion
            if a in self._users_map:
                sender = self._users_map[a]['email']
            else:
                sender = self.zulip.get_profile()['email']
```

hcraT replied:
I'm aware of the double call. I generally prefer to have a default value and only one branch of the if. In this case I can see that I do an API call and so this is not good. I'll cache the email in the constructor so that everything will be local and no unnecessary calls will be made

tmichela reviewed (changes requested):
few minor comments

hcraT committed:
Fixes the issues reported in the review.

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -4,7 +4,11 @@
 import warnings
+
+import csv
+import datetime
```
```suggestion
import csv
import datetime
import os
import re
import sys
import warnings
```

Let's have a single group, since they're all in the standard library

tmichela commented on elog_zulip/__init__.py:
```diff
@@ -53,23 +57,84 @@ class Elog:
+                    if not can_create_users:
+
+                        response = self.zulip.get_members()
+
+                        if response['result'] == 'success':
+                            existing_users = { m['email'] : m for m in response['members'] }
+                        else:
+                            log.error(f'Error while querying the users: {response["msg"]}')
+                            sys.exit(1)
+
+
+                    with open(users_map_path, newline='') as f:
+
+                        data = list(csv.reader(f))
+                        h = data[0]
+                        self._users_map = {}
+
+                        if len(data) < 2:
+                            log.error(f'The provided user map is empty. Please provide a valid one.')
+                            sys.exit(1)
+
+                        if not ('elog_user' in h and 'email' in h):
+                            log.error(f'The users map needs to have an header with two fields: "elog_user" (name of the user in the elog) and "email" (email of the user in the destination zulip system)')
+                            sys.exit(1)
+
+                        for d in data[1:]:
+                            user_dict = dict(zip(h, d))
+
+                            if not can_create_users and user_dict['email'] not in existing_users:
+                                log.info(f'User {user_dict["email"]} is not present in the server skipping.')
+                                continue
+
+                            self._users_map[user_dict['elog_user']] = user_dict
+
+                        if len(self._users_map) == 0:
+                            log.error(f'No user found in the user map. Please check that you are using the correct user map for this elog.')
+                            sys.exit(1)
+                else:
+                    log.error(f'Cannot proceed without a users map')
+                    sys.exit(1)
```
Just suggesting minor refactoring, Feel free to ignore though.
- move the initialization of the user map in a separate method
- reducing the depth of blocks

logic should not have changed (just moved one data access after checking the data list has data).

```suggestion
        self._fallback_user = self.zulip.get_profile()['email']

        self._users_map = {}
        if self.impersonate:
            if not users_map_path:
                log.error(f'Cannot proceed without a users map')
                sys.exit(1)
            self._load_elog_user_map()

    def _load_elog_user_map(self):
        can_create_users = self.config.get('can-create-users', False)
        existing_users = {}

        if not can_create_users:
            response = self.zulip.get_members()
            if response['result'] == 'success':
                existing_users = { m['email'] : m for m in response['members'] }
            else:
                log.error(f'Error while querying the users: {response["msg"]}')
                sys.exit(1)

        with open(self.config.get('user-map'), newline='') as f:
            data = list(csv.reader(f))

        if len(data) < 2:
            log.error(f'The provided user map is empty. Please provide a valid one.')
            sys.exit(1)

        header = data[0]

        if not ('elog_user' in header and 'email' in header):
            log.error(
                f'The users map needs to have an header with two fields:'
                '"elog_user" (name of the user in the elog) and "email" (email of '
                 'the user in the destination zulip system)')
            sys.exit(1)

        for d in data[1:]:
            user_dict = dict(zip(header, d))

            if not can_create_users and user_dict['email'] not in existing_users:
                log.info(f'User {user_dict["email"]} is not present in the server skipping.')
                continue

            self._users_map[user_dict['elog_user']] = user_dict

        if len(self._users_map) == 0:
            log.error(f'No user found in the user map. Please check that you are using the correct user map for this elog.')
            sys.exit(1)
```

tmichela reviewed (commented):
Generally LGTM,
Just couple minor suggestions

hcraT committed:
Update elog_zulip/__init__.py

Move all std library dependencies in a single alphabetically ordered group.

Co-authored-by: Thomas Michelat <32831491+tmichela@users.noreply.github.com>

hcraT committed:
Update elog_zulip/__init__.py

Improve code readability by refactoring code into a function and removing unnecessary nesting.

Co-authored-by: Thomas Michelat <32831491+tmichela@users.noreply.github.com>

hcraT commented:
I'll do a new test just to be sure that everything works and then I think we can merge.

hcraT committed:
Fixes wrong configuration path

hcraT commented:
I succesfully did an import. So if you agree LGTM.

Thank you for your review!

hcraT commented:
I found a small issue while doing some tests and added a commit.

hcraT committed:
Allows to specify a default user for impersonation in the configuration.

hcraT commented:
@tmichela Given the new changes are not minimal this will probably need another round of review.
